src/Tools/CsvImporter/Services/MainService/MainService.cs

```diff
@@ -0,0 +1,291 @@
+            List<UserDetails> filteredUserDetailsList = userDetailsList.FindAll(
+                match: userDetails => userDetails.SecondaryEmail is not null || userDetails.PhoneNumber is not null
+            );
+
+            _logger.LogInformation("Filtered to {FilteredUserDetailsCount} users with email or phone number", filteredUserDetailsList.Count);
+
+            // If there are no users to process, exit.
+            if (filteredUserDetailsList.Count == 0)
+            {
+                _logger.LogWarning("No users to process, exiting");
+                exitCode = 10;
+
+                return;
+            }
+
+            // If this is a dry run, exit.
+            if (_configService.Config.DryRunEnabled)
+            {
+                _logger.LogWarning("Dry run, exiting");
+                exitCode = 5;
+
+                return;
+            }
+
+            // Set the initial semaphore count to half the max tasks.
+            double initialTasksCount = Math.Round((double)(maxTasks / 2), 0);
+
+            using SemaphoreSlim semaphoreSlim = new(
+                initialCount: (int)initialTasksCount,
+                maxCount: maxTasks
+            );
+
+            // Process each item and send to queue.
+            List<Task> tasks = [];
+            foreach (var userItem in filteredUserDetailsList)
+            {
+                UserAuthUpdateQueueItem queueItem = new()
+                {
+                    EmployeeId = userItem.EmployeeNumber,
+                    UserName = userItem.UserName,
+                    EmailAddress = userItem.SecondaryEmail,
+                    PhoneNumber = userItem.PhoneNumber
+                };
+
+                var newQueueItemTask = SendUserAuthUpdateQueueItemAsync(semaphoreSlim, queueItem);
+
+                tasks.Add(newQueueItemTask);
+            }
+
+            // Wait for all tasks to complete.
+            _logger.LogInformation("Waiting for tasks to complete...");
+            await Task.WhenAll(tasks);
+
+            // Copy the CSV file used for this run to the config directory.
+            _logger.LogInformation("Saving last run CSV file path to config file.");
+            string copiedCsvFilePath = Path.Combine(_configService.GetConfigDirPath(), "lastRun.csv");
+            File.Copy(
+                sourceFileName: csvFileInfo.FullName,
+                destFileName: copiedCsvFilePath,
+                overwrite: true
+            );
+
+            // Update the config file.
+            _configService.Config.LastCsvPath = copiedCsvFilePath;
+            _configService.Config.LastRunDateTime = DateTimeOffset.UtcNow;
+            await _configService.SaveConfigAsync();
+
+            exitCode = 0;
+            return;
+        }
+        finally
+        {
+            stopwatch.Stop();
+
+            _logger.LogInformation("Completed in {ElapsedMilliseconds}ms", stopwatch.ElapsedMilliseconds);
+
+            Environment.Exit(exitCode);
```

Comment:
This could be problematic, [see this](https://learn.microsoft.com/en-us/dotnet/core/extensions/generic-host?tabs=appbuilder#host-shutdown).